### PR TITLE
Allow To Not Overwrite State By ID

### PR DESCRIPTION
**Pull request by austinkahly**
Now that the tests are passing and the code is in the right spot, we can review this and start testing it out. :wink:

### CODE REVIEW

File: lib/ar_state_machine.rb [hunk 1]
```diff
@@ -121,6 +121,19 @@ def allow_transition?(states, from, to)
     states[from.to_sym].include?(to.to_sym)
   end
 
+  def set_state_by_id
```
Comment (ksolkowski) [inline lib/ar_state_machine.rb]:
```ruby
def set_state_by_id(ss)
  if self.respond_to?("#{ss}_by_id")
    overwrite = true
    if self.respond_to?("overwrite_#{ss}_by_id")
      overwrite = !(self.send("overwrite_#{ss}_by_id") == false)
    elsif self.class.respond_to?("overwrite_#{ss}_by_id")
      overwrite = !(self.class.send("overwrite_#{ss}_by_id") == false)
    end
    if self.send("#{ss}_by_id").blank? or overwrite
      self.send("#{ss}_by_id=", self.last_edited_by_id)
    end
  end
end
```

then in the make_state functions... just need to do `set_state_by_id(ss)` after it sets self.last_edited_by_id

File: lib/ar_state_machine.rb [hunk 2]
```diff
@@ -179,10 +194,8 @@ def setup(states)
         define_method "make_#{ss}" do |last_edited_by_override = nil|
           self.state = ss
           self.last_edited_by_id = last_edited_by_override if last_edited_by_override
+          self.set_state_by_id()
```
Comment (fiestacasey) [inline lib/ar_state_machine.rb]: () is not necessary. this is ruby

Comment (austinkahly) [reply lib/ar_state_machine.rb]: kevin and I both felt it looked really weird without the (), but if you want it, I'll remove it.

File: spec/lib/state_machine_spec.rb [hunk 3]
```diff
@@ -126,10 +126,25 @@
...
   end
+
+  it "test overwriting ids" do
+    test = StateMachineTestClass.create
+    expect(test.second_state_by_id).to be_nil
+
+    expect(test.make_second_state(2)).to be true
+
+    was = test.second_state_by_id
+    test.overwrite_second_state_by_id = false
```
Comment (ksolkowski) [inline spec/lib/state_machine_spec.rb]: perhaps set it up like we do in wantable-2
```ruby
  cattr_accessor :overwrite_billing_failed_at
  self.overwrite_billing_failed_at = false
```
?

Comment (austinkahly) [reply spec/lib/state_machine_spec.rb]: I set this test up like we did for the overwrite_second_state_at, which is working. Are you cool with that? or do you want me to change it all up?

Comment (ksolkowski) [reply spec/lib/state_machine_spec.rb]: nah. w/e it cool

**Pull request merged**